ADR-019/020 cleanups: HybridCodec, Pubkey (#8683)
cosmos/cosmos-sdk: docs/architecture/adr-019-protobuf-state-encoding.md

persisted structured data in the Cosmos SDK while providing a clean mechanism and developer UX for
applications wishing to continue to use Amino. We will provide this mechanism by updating modules to
accept a codec interface, `Marshaler`, instead of a concrete Amino codec. Furthermore, the Cosmos SDK
will provide two concrete implementations of the `Marshaler` interface: `AminoCodec` and `ProtoCodec`.

- `AminoCodec`: Uses Amino for both binary and JSON encoding.
- `ProtoCodec`: Uses Protobuf for both binary and JSON encoding.

Modules will use whichever codec that is instantiated in the app. By default, the SDK's `simapp`
instantiates a `ProtoCodec` as the concrete implementation of `Marshaler`, inside the `MakeTestEncodingConfig`
function. This can be easily overwritten by app developers if they so desire.

The ultimate goal will be to replace Amino JSON encoding with Protobuf encoding and thus have

> modules accept and/or extend `ProtoCodec`. Until then, Amino JSON is still provided as a fallback.

Reviewer: ```suggestion
modules accept and/or extend `ProtoCodec`. Until then, Amino JSON is still provided for legacy use-cases.
```
